src/k8s/pkg/k8sd/app/cluster_util.go

```diff
@@ -0,0 +1,131 @@
+package app
+
+import (
+	"fmt"
+	"net"
+	"path"
+
+	"github.com/canonical/k8s/pkg/k8sd/api/impl"
+	"github.com/canonical/k8s/pkg/k8sd/pki"
+	"github.com/canonical/k8s/pkg/k8sd/setup"
+	"github.com/canonical/k8s/pkg/k8sd/types"
+	"github.com/canonical/k8s/pkg/snap"
+	snaputil "github.com/canonical/k8s/pkg/snap/util"
+	"github.com/canonical/k8s/pkg/utils/k8s"
+	"github.com/canonical/microcluster/state"
+)
+
+func generateAndEnsureCertificates(snap snap.Snap, cfg types.ClusterConfig, nodeName string, allowSelfSignedCA bool) error {
+	// Certificates
+	switch cfg.APIServer.Datastore {
+	case "k8s-dqlite":
+		certificates := pki.NewK8sDqlitePKI(pki.K8sDqlitePKIOpts{
+			Hostname:          nodeName,
+			IPSANs:            []net.IP{{127, 0, 0, 1}},
+			Years:             20,
+			AllowSelfSignedCA: allowSelfSignedCA,
+		})
+		certificates.K8sDqliteCert = cfg.Certificates.K8sDqliteCert
+		certificates.K8sDqliteKey = cfg.Certificates.K8sDqliteKey
+		if err := certificates.CompleteCertificates(); err != nil {
+			return fmt.Errorf("failed to initialize cluster certificates: %w", err)
+		}
+		if err := setup.EnsureK8sDqlitePKI(snap, certificates); err != nil {
+			return fmt.Errorf("failed to write cluster certificates: %w", err)
+		}
+	case "external":
+		certificates := &pki.ExternalDatastorePKI{
+			DatastoreCACert:     cfg.Certificates.DatastoreCACert,
+			DatastoreClientCert: cfg.Certificates.DatastoreClientCert,
+			DatastoreClientKey:  cfg.Certificates.DatastoreClientKey,
+		}
+		if err := certificates.CheckCertificates(); err != nil {
+			return fmt.Errorf("failed to initialize cluster certificates: %w", err)
+		}
+		if err := setup.EnsureExtDatastorePKI(snap, certificates); err != nil {
+			return fmt.Errorf("failed to write cluster certificates: %w", err)
+		}
+	default:
+		return fmt.Errorf("unsupported datastore %s, must be one of %v", cfg.APIServer.Datastore, setup.SupportedDatastores)
+	}
+
+	return nil
+
+}
+
+func generateKubeconfigs(snap snap.Snap, s *state.State, cfg types.ClusterConfig) error {
```

Comment:
Reduce what is passed here, and adjust name to better reflect what the function does without having to lookup the implementation
```suggestion
func setupKubeconfigs(ctx context.Context, snap snap.Snap, securePort int, caCert string) error {
```

Reply:
Couldn't get rid of passing state here but I also reduced what gets passed for the other functions.